qiskit/circuit/quantumcircuit.py

```diff
@@ -477,22 +467,44 @@ def extend(self, rhs):
         # Check registers in LHS are compatible with RHS
         self._check_compatible_regs(rhs)
 
-        # Add new registers
-        for element in rhs.qregs:
-            if element not in self.qregs:
-                self.qregs.append(element)
-        for element in rhs.cregs:
-            if element not in self.cregs:
-                self.cregs.append(element)
-
-        # Copy the circuit data if rhs and self are the same, otherwise the data of rhs is
-        # appended to both self and rhs resulting in an infinite loop
-        data = rhs.data.copy() if rhs is self else rhs.data
-
-        # Add new gates
-        for instruction_context in data:
-            self._append(*instruction_context)
-        return self
+        warnings.warn("The QuantumCircuit.extend() method is being deprecated."
+                      "Use the compose() method which is more flexible w.r.t "
+                      "circuit register compatibility.", DeprecationWarning)
+        return self.compose(rhs)
+
+    def compose(self, other, qubits=None, clbits=None, front=False, inplace=False):
```

Comment:
`compose` was already defined in https://github.com/Qiskit/qiskit-terra/pull/4208/files#diff-671fa09c49f08cbefb5f78fd69b0b6bbR509